### ISSUE TITLE
Insert plot level height histogram into Clowder geostreams; height into BETYdb

### ISSUE THREAD

**Issue by ZongyangLi** [terraref/computing-pipeline#210]
### Description

We have scripts to generate plot level height histogram on Roger. The next step is to create a pipeline for this extractor.

### Completion Criteria

- [x] @solmazhajmohammadi need to provide a parameter of "the position of the calibration object"
    - we are now making our assumptions of where is the (0,0,0) point in the field coordinate system.
    - from @solmazhajmohammadi we know that "We have used a special object to calibrate the point clouds. (0,0,0) point in the point cloud is somewhere in middle of the field, we will provide the transformation matrix to the gantry coordinate system in metadata"
- [x] make sure the correlation between hand measurements and predictions is high enough
    - correlation in 9/8, 9/15, 9/22, 9/29 is around 90% with our assumptions
    - we found that in 8/31, almost all of plots have a same highest height level at lv. 83, it seems point above lv. 83 was 'disappeared'.
- [ ] add to pipeline
    - insert histogram into geostreams database as attribute
    - need a set of raw data from clowder as input, like #85 
    - insert estimate of canopy_height into BETYdb with appropriate method

**Comment by dlebauer** [comment]
@ZongyangLi @rmgarnett @pless 


For this extractor, I would suggest that we write the summary stats (histogram) into the metadata and insert a few statistics into BETYdb. For example, we have inserted a trait called '95th quantile height'.

But the key trait from the point cloud is the height estimate calibrated to field measurements. This trait will have the same name as the trait that Maria measured, i.e. 'canopy_height'. I think it would make sense for this extractor to use the calibrated model that Roman developed in #175.

**Comment by dlebauer** [comment]
@rmgarnett what are the (slope, intercept) parameters from the model in #175?

When estimating height at the plot level, can we also estimate uncertainty?

**Comment by rmgarnett** [comment]
[hand height] = 28.2cm + 0.661 * [89th height percentile]

The RMSE/MAE gives a rough estimate of L2/L1 uncertainty. I will do a more thorough analysis in January now that all height distributions are extracted.

**Comment by dlebauer** [comment]
@rmgarnett I suspect RMSE scales with height?

From your plot it is hard to tell how the data are distributed b/c of overlapping points. But I gather strongly right-skewed. I wonder if log transforming x and y would be appropriae, if it would more evenly weight the smaller values. The small plants are important too!

![img_2204](https://cloud.githubusercontent.com/assets/464871/21432070/c5f5cadc-c81e-11e6-9fa7-fa78c9eb6336.PNG)

**Comment by ZongyangLi** [comment]
@dlebauer 
I have got all height distribution data for season 2 from 8/8 to 11/25, and I created 90th and 95th height percentile csv file, according to @rmgarnett 's research. 
[90th percentile](https://drive.google.com/open?id=0B5QCp_Onc6nOTi1MdGhiMlZ4VkU)
[95th percentile](https://drive.google.com/open?id=0B5QCp_Onc6nOZmE5MUt5d1JwSnM)
Scanner3DTop data in Season 2 is much better than those in Season 1, but still data from 10/13 to 11/04 are unexpected, there are just a few points in those days ply files.

I am wondering if point cloud files might be fixed in those days, if not, what's your opinion of putting them into BETYdb.

**Comment by dlebauer** [comment]
@solmazhajmohammadi could you please check into whether we can recover useful data from 10/13 to 11/04?

@ZongyangLi we need to discuss with @rmgarnett about how to implement this extractor.

**Comment by dlebauer** [comment]
@rmgarnett have you made any progress on adding uncertainty?

**Comment by rmgarnett** [comment]
I will pick this up again this week.

**Comment by dlebauer** [comment]
@ZongyangLi you can go ahead and insert the data that you have. We can create another issue for adding uncertainty to the height calculations (moving forward this should be done by default ... )

**Comment by solmazhajmohammadi** [comment]
@dlebauer @ZongyangLi, for the data from 10/13 to 11/04, the png files have not been collected correctly, but we can get the height information from the scans that it is done at ~5m

**Comment by solmazhajmohammadi** [comment]
@smarshall-bmr can you please scan the checker boards to find the pointcloud origin?

**Comment by ZongyangLi** [comment]
@solmazhajmohammadi,  are you saying to estimate the plot level height base on the highest points in the remaining 3d data? That might be different from what we have done before, because we are using all point cloud data to create a height histogram and calculate quantiles data to make predictions.

**Comment by solmazhajmohammadi** [comment]
@ZongyangLi This could be an option, otherwise the data has been collected with a wrong setting, so we are not able to recover it.

**Comment by rmgarnett** [comment]
I have been reinvestigating the hand measurements using @ZongyangLi's most-recent data. The final model may differ from what's written above, but it will be the same form. I presume the extractor will be easy to modify if we wish to change the model slightly?

**Comment by dlebauer** [comment]
Yes, we could store parameters as metadata and have the extractor pick them
up (eg if they change by crop, year, or location)

**Comment by rmgarnett** [comment]
Perfect.

**Comment by rachelshekar** [comment]
@robkooper and @max-zilla - should this go into geostreams?  @ZongyangLi -what will be visualized?

**Comment by solmazhajmohammadi** [comment]
@rmgarnett @ZongyangLi 
PointCloud data from 2016/08/07 to 2016/09/05 was collected with a wrong setting. There is no way to fix this dataset. Maybe we can delete them or mark them to exclude from the pipeline.
@dlebauer @max-zilla any idea?

**Comment by dlebauer** [comment]
Please don't remove them unless it is clear that they contain no useful
information - i.e. all points have been randomly redistributed. We can keep
them but exclude them from our workflow (even if the information is not
useful within the current scope of the project, others may find it useful).

One possibility:
* Describe the problem in a file called 'readme-error' and put it in every
folder with unuseable data.
* Append all files with -error
* Have extractors skip files containing 'error' in the filename or
directory path

**Comment by rmgarnett** [comment]
I think that's reasonable.  Unfortunately, I am not sure these points can be used to reliably estimate height, but they could be useful for some other purpose.

I could try to learn a separate model for this range and for the days afterwards, but I hesitate to do so.

**Comment by max-zilla** [comment]
@ZongyangLi @rmgarnett any updates on this issue?

**Comment by ZongyangLi** [comment]
I was worried about providing untrue data into Clowder geostream or BETYdb, so I asked for a transformation matrix to the gantry coordinate system in metadata to get a plot level height histogram. 
If these uncertain result can be insert into those database, @max-zilla could you send me an instruction of Clowder geostreams?

**Comment by max-zilla** [comment]
@ZongyangLi we can write data to geostreams that we can regenerate and replace later once we have corrections to the code, absolutely.

Here is a comment I left in another issue about Geostreams:
https://github.com/terraref/computing-pipeline/issues/252#issuecomment-286189327

If you look at the links there, you can see an example of how I use it. The basic approach is:

1) determine which "plot" /sensor to use. I've already created a geostreams sensor entry for each plot, and you can query for the nearest one by lat/long with this:
https://github.com/terraref/extractors-metadata/blob/use-pyclowder-geostreams/sensorposition/terra_sensorposition.py#L101
```
sensor_data = pyclowder.geostreams.get_sensors_by_circle(connector, host, secret_key, sensor_latlon[1], sensor_latlon[0], 0.01)
```

2) determine which stream to use. you'll want to create a new stream for your data within the plot. e.g. "Height Histogram - Range X Pass Y" (where Range X Pass Y is the name of returned in sensor_data above). Here is a code snippet where we can look for an existing stream with that name and create it if it doesn't exist:
https://github.com/terraref/extractors-metadata/blob/use-pyclowder-geostreams/sensorposition/terra_sensorposition.py#L129
```
stream_data = pyclowder.geostreams.get_stream_by_name(connector, host, secret_key, stream_name)
		if not stream_data:
			stream_id = pyclowder.geostreams.create_stream(connector, host, secret_key, stream_name, sensor_id, {
				"type": "Point",
				"coordinates": [sensor_latlon[1], sensor_latlon[0], 0]
			})
		else: stream_id = stream_data['id']
```

3) Add datapoints to that stream ID.
https://github.com/terraref/extractors-metadata/blob/use-pyclowder-geostreams/sensorposition/terra_sensorposition.py#L161
The "metadata" JSON properties object can have whatever you want - at least the height histogram in this case.

Take a look at the code and let me know if it kind of makes sense. If it's helpful, here's the pyclowder 2 geostreams source code:
https://opensource.ncsa.illinois.edu/bitbucket/projects/CATS/repos/pyclowder2/browse/pyclowder/geostreams.py

**Comment by ZongyangLi** [comment]
@max-zilla Thanks a lot! I will take a look into it and start working on this.

**Comment by max-zilla** [comment]
@ZongyangLi the "error registering extractors" is not a big problem - registering just makes Clowder allow that extractor to be selected in manual extractor lists in the GUI. If you got "Waiting for messages" I think it's working properly. 

you do want pyclowder 2, yes

**Comment by ZongyangLi** [comment]
@max-zilla 
Could you give me the definition of all the input arguments to the geostreams. Because it seems I need to build all the arguments myself, such as `sensor_data stream_name geom`and so on

**Comment by max-zilla** [comment]
@ZongyangLi that is included in the geostreams source code:
https://opensource.ncsa.illinois.edu/bitbucket/projects/CATS/repos/pyclowder2/browse/pyclowder/geostreams.py

```
def create_sensor(connector, host, key, sensorname, geom, type, region):
    """Create a new sensor in Geostreams.

    Keyword arguments:
    connector -- connector information, used to get missing parameters and send status updates
    host -- the clowder host, including http and port, should end with a /
    key -- the secret key to login to clowder
    sensorname -- name of new sensor to create
    geom -- GeoJSON object of sensor geometry
    type -- JSON object with {"id", "title", and "sensorType"}
    region -- region of sensor
    """

def create_stream(connector, host, key, streamname, sensorid, geom, properties={}):
    """Create a new stream in Geostreams.

    Keyword arguments:
    connector -- connector information, used to get missing parameters and send status updates
    host -- the clowder host, including http and port, should end with a /
    key -- the secret key to login to clowder
    streamname -- name of new stream to create
    sensorid -- id of sensor to attach stream to
    geom -- GeoJSON object of sensor geometry
    properties -- JSON object with any desired properties
    """

def create_datapoint(connector, host, key, streamid, geom, starttime, endtime, properties={}):
    """Create a new datapoint in Geostreams.

    Keyword arguments:
    connector -- connector information, used to get missing parameters and send status updates
    host -- the clowder host, including http and port, should end with a /
    key -- the secret key to login to clowder
    streamid -- id of stream to attach datapoint to
    geom -- GeoJSON object of sensor geometry
    starttime -- start time, in format 2017-01-25T09:33:02-06:00
    endtime -- end time, in format 2017-01-25T09:33:02-06:00
    properties -- JSON object with any desired properties
    """

def get_sensor_by_name(connector, host, key, sensorname):
    """Get sensor by name from Geostreams, or return None.

    Keyword arguments:
    connector -- connector information, used to get missing parameters and send status updates
    host -- the clowder host, including http and port, should end with a /
    key -- the secret key to login to clowder
    sensorname -- name of sensor to search for
    """

def get_sensors_by_circle(connector, host, key, lon, lat, radius=0):
    """Get sensor by coordinate from Geostreams, or return None.

    Keyword arguments:
    connector -- connector information, used to get missing parameters and send status updates
    host -- the clowder host, including http and port, should end with a /
    key -- the secret key to login to clowder
    lon -- longitude of point
    lat -- latitude of point
    radius -- distance in meters around point to search
    """
```

As an aside, not sure how you're chopping these to plots right now but this task becomes a lot easier after we have a way to stitch + clip images to plots: https://github.com/terraref/computing-pipeline/issues/265

**Comment by dlebauer** [comment]
@ZongyangLi is this extractor ready to deploy?

**Comment by ZongyangLi** [comment]
@dlebauer 
Insert existed 'height' trait into BETYdb is ready, the only thing I need your confirm is using 864 plots or 1728 plots.
To make it as an extractor in clowder, I still need some support mentioned here:
https://github.com/terraref/computing-pipeline/issues/193#issuecomment-290507753

**Comment by max-zilla** [comment]
@ZongyangLi sorry if I missed it, where is code for this extractor? I know the canopycover extractor is in https://github.com/terraref/extractors-stereo-rgb

If you can share what you have I can contribute to the Clowder part.

**Comment by ZongyangLi** [comment]
@max-zilla Insert existed 'height' trait into BETYdb is not an extractor, 'height' data for season 2 is on my desktop, I have a local script can upload them to BETYdb.

To make it as an extractor in clowder there are something I am not sure:
1. (0,0,0) point in the point cloud is somewhere in middle of the field, I have no idea where is this (0,0,0) point in gantry coordinate system, and what's the positive orientation for west scan and east scan, this relate to https://github.com/terraref/reference-data/issues/44.
2. ‘an individual plot is made up of several ply files, which means I need a set of raw data from clowder as input automatically’

I made a comment here https://github.com/terraref/computing-pipeline/issues/193#issuecomment-270767586 to describe the support I needed and discuss with @dlebauer , but I didn't find the answer to this issue yet. I am sorry too if I missed any response for this.

**Comment by max-zilla** [comment]
@ZongyangLi for #2, unless @dlebauer says different I say we do your second option:

Another way of solving this problem is create several traits for a plot for one day if it is acceptable, and that will be much easier.

...when the field stitching is done, we can modify this extractor to trigger on the full merged PLY file and avoid the need to generate multiple traits, but we can do that for now just to have something running.

I pinged Solmaz and Stuart about #1.

**Comment by ZongyangLi** [comment]
@max-zilla 
This comment is from @dlebauer 'While It is possible to add replicate measurements for a single plot (we are already doing this for field data), we should do this when it is scientifically useful and not because of the way that the data are written.'

**Comment by max-zilla** [comment]
PLY chunks...
1) need to consider sampling rates across PLY files so we dont mix and match. 
2) better to stitch & subset according to a design. 
3) we can deploy what we have now & replace with final version by end of May.

**Comment by ZongyangLi** [comment]
@max-zilla I am going to upload my code associate with plant height. Is there anywhere on Github may I upload and update my code?

**Comment by max-zilla** [comment]
@ZongyangLi if using PLY data, please upload here: https://github.com/terraref/extractors-3dscanner

**Comment by ZongyangLi** [comment]
@max-zilla 
Okay! Could you please create a new directory there?

**Comment by max-zilla** [comment]
@ZongyangLi created 'plantheight' directory

**Comment by dlebauer** [comment]
I don't see the plantheight directory, but could you change it to
plant_height to match the naming in BETYdb?

**Comment by ZongyangLi** [comment]
codes updated to https://github.com/terraref/extractors-3dscanner/tree/master/plantheight

**Comment by max-zilla** [comment]
@dlebauer @ZongyangLi renamed to plant_height:
https://github.com/terraref/extractors-3dscanner/tree/master/plant_height

**Comment by max-zilla** [comment]
created this issue to finish deployment: https://github.com/terraref/computing-pipeline/issues/303

There was an error scanning the PLY, but code is 90% there. Last missing piece is how to extract the actual values to geostreams - I can create the datapoints if you can show how to get an array of values from the histogram.npy file

**Comment by max-zilla** [comment]
@ZongyangLi will mention on phone, but remaining piece is to extract values from height file for BETYdb and histogram file for geostreams. Right now these outputs are just uploaded to Clowder as .npy files.

**Comment by ZongyangLi** [comment]
@max-zilla I think there are still some other issues on this extractor. Maybe we could ignore them for now, but it will not be a true pip-line without solving these problem

1. The single scan covers about 1 meters in NS-direction, a plot length in NS is around 4 meters, we will need to integrate 4 or 5 scan's data to get a plot-level height.
2. The existed 'yshift' value here https://github.com/terraref/extractors-3dscanner/blob/master/plant_height/full_day_to_histogram.py#L597 only works for season2. This relate to https://github.com/terraref/reference-data/issues/44 as I mentioned before.
3. To extract values from histogram.npy, they are here:
https://github.com/terraref/extractors-3dscanner/blob/master/plant_height/full_day_to_histogram.py#L277

**Comment by max-zilla** [comment]
@solmazhajmohammadi is going to share the point cloud offset to gantry for #2.

for #1, calibrated in august and validation in November. Saw misalignment between two different point clouds ~4cm. Suspect that alignment is shifting with temperature. @smarshall-bmr will do that scan. maximum we recorded up to now is about ~5cm between hot summer and middle winter.

**Comment by max-zilla** [comment]
@ZongyangLi will look at 2017 data for the merge and ignore season 1 issues for the moment.

I will add the geostreams function and then we can deploy initial version.

**Comment by solmazhajmohammadi** [comment]
@smarshall-bmr it seems that the misalignment in pointclouds is due to the temperature change. It would be great if you can do multiple scan in different time of the day to see the variation with changing the temperature. 
Thanks

**Comment by ZongyangLi** [comment]
@dlebauer I'm going to insert height data into BETYdb. Here is an example of csv file. Could you please review and have a check. Thanks!

https://drive.google.com/open?id=0B5QCp_Onc6nOUGR0WVpsREl5NWc

**Comment by dlebauer** [comment]
@zongyangli that looks good. Go ahead and upload. Thanks!

**Comment by ZongyangLi** [comment]
@dlebauer 
When I do that, I receive below error report:
`"mean": "31.505",
            "local_datetime": "2016-08-08T12:00:00",
            "access_level": "2",
            "error": "bad date specification; see error output",
            "species": {
              "scientificname": "Sorghum bicolor"
            },`
`"site": {
              "sitename": "MAC Field Scanner Field Plot 1603 Season 2",
              "error": "match not found"
            },`
`"method": {
              "name": "height Estimation from 3D Scanner using formula: [hand height] = 28.2cm + 0.661 * [89th height percentile]",
              "error": "match not found"
            },`

**Comment by dlebauer** [comment]
@zongyangli 

1. The date looks correct - what is the error output?
2. The sitename needs to match a record in the database terraref.ncsa.illinois.edu/bety/sites
2. the method name needs to match a name in the methods table terraref.ncsa.illinois.edu/bety/methods

**Comment by ZongyangLi** [comment]
@dlebauer 
1. the error is 'bad data specification', I think it's because of error 2 and error 3.
2. This is a plot number definition problem I mentioned here: https://github.com/terraref/computing-pipeline/issues/210#issuecomment-295475495, here I upload a totally 1728 plots data, maybe this makes error 2.
3. the method name comes from you comment here: https://github.com/terraref/computing-pipeline/issues/193#issuecomment-271668626, I failed to access the methods table.

**Comment by dlebauer** [comment]
1. Hmm. error 1 is the only one that says 'see error output'. @gsrohde any ideas?
2. For season 1, Plot names for the 1728 plots are the same as for the 864 plots, but with E, W appended. We don't have the same 1728 plots specified for season 2; I can create the 1728 plots for Season 2 or we could go with the 864 plots. 
3. sorry I was not clear. for the method, please create a new record here: https://terraref.ncsa.illinois.edu/bety/methods/new. Can you access that page?

**Comment by ZongyangLi** [comment]
@dlebauer I successfully create a new method called: 'Scanner 3d ply data to height', and error 3 goes away. Since the 1728 plots still not work, I used the 864 plots definition to finish the operation.

One another error comes out:
`"model_validation_errors": "{:mean=>[\"The value of mean for the height trait must be at most 120.\"]}",`

By the way, @solmazhajmohammadi gave us the origin in point cloud data last week, I am creating new height estimate data using the new metadata. 

Also, we are creating algorithm to use stereo top image to do the 3d reconstruction, and trying to use this stereo 3d model to recover the missing height data from 2016/10/11 to 2016/11/04, once we get the new available data, I will resend them to BETYdb.

**Comment by dlebauer** [comment]
@ZongyangLi thanks,

* the error you observed is caused by the fact that each variable has a valid range, and in the case of 'height' the valid range is [0, 120]. And the variable height has units in meters (as you can tell, these ranges are generally very generous; the tallest known tree is 115m). 
* however, I think that we should be using the variable [`canopy_height](https://terraref.ncsa.illinois.edu/bety/traits/6000000346), since this is the variable that Maria recorded, and I assume that these are the data used to fit the model (please confirm) 
* See also https://github.com/terraref/reference-data/issues/118 for disambiguating terms related to height.

In summary:
* if the model is trained only on Maria's canopy_height measurements, we should call the output canopy_height.
* if it is trained on both canopy_height and either panicle_height or its synonym spike_height, we should call the variable plant_height (plant_height = max(canopy_height, panicle_height). 
* I am assuming that the algorithm does not differentiate plants with or without fruit.

**Comment by gsrohde** [comment]
@dlebauer and @ZongyangLi The message should probably say "See error list" instead.  The error list contains the key-value pair `date_data_errors=>[\"You can't have a local_datetime attribute on a trait-group's defaults element if no default site has been specified.\"]`.  But this itself is unclear except when an XML file, not a CSV file, is being uploaded.  And I think the error itself is bogus in this case.  I'm making an issue for this.

**Comment by dlebauer** [comment]
* change to plant_height

**Comment by rachelshekar** [comment]
Currently this extractor generates two numpy files:
```
ls /projects/arpae/terraref/sites/ua-mac/Level_1/scanner3DTop_plant_height/2017-05-06/2017-05-06__05-05-58-934
scanner3DTop - 2017-05-06__05-05-58-934 highest.npy  scanner3DTop - 2017-05-06__05-05-58-934 histogram.npy
```

In #303 we mention pushing the max value to geostreams - this isn't implemented yet. 

Additionally we should convert the .npy files to an image or geotiff.

**Comment by ZongyangLi** [comment]
@max-zilla I have just uploaded recent codes to https://github.com/terraref/extractors-3dscanner/tree/master/plant_height
And I know it's really a mess code, I need to take some time to make it much more beautiful.

**Comment by max-zilla** [comment]
@ZongyangLi I ran a PLY file from feb through the plant_height extractor - I'm making a pull request to show @nickheyek that writes .tif instead of .npy and pushes some values to geostreams.

The "highest" numpy array looks like this for sample file:
```
>>> f = r"/Users/mburnette/globus/111fb573-a351-4bf8-9594-aedddc53a850__Top-heading-east_0.ply"                       
>>> plydata = PlyData.read(f)
>>> hist, highest = full_day_to_histogram.gen_height_histogram(plydata, False)
>>> highest
array([[ 1051.20507812],
       [ 1042.48278809],
       [ 1005.7802124 ],
       [ 1016.98535156],
       [ 1013.16113281],
       [ 1053.45422363],
       [ 1031.78356934],
       [ 1024.53540039],
       [ 1019.98809814],
       [ 1020.35986328],
       [ 1007.30383301],
       [  948.19873047],
       [  958.54162598],
       [  989.56195068],
       [    0.        ],
       [    0.        ]])
>>> len(highest)
16
```

What exactly do these 16 numbers represent? I see the "hist" array is also length 16, but each of its members is itself an array:
```
>>> hist
array([[ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       ..., 
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.]])
>>> len(hist)
16
```

If I were to summarize the "highest" numbers into geostreams, what would you recommend? Something like this is hard to interpret:
```
{
"highest": [[ 1051.20507812],
       [ 1042.48278809],
       [ 1005.7802124 ],
       [ 1016.98535156],
       [ 1013.16113281],
       [ 1053.45422363],
       [ 1031.78356934],
       [ 1024.53540039],
       [ 1019.98809814],
       [ 1020.35986328],
       [ 1007.30383301],
       [  948.19873047],
       [  958.54162598],
       [  989.56195068],
       [    0.        ],
       [    0.        ]]
}
```

Can I select the maximum number of this array as a simple "tallest plant in this plot" number?

**Comment by ZongyangLi** [comment]
- What exactly do these 16 numbers represent?
Every ply file supposes to cover one row between east boundary and west boundary in the field. At the very beginning, field were divided into 16 column, these 16 numbers represent for the highest point in the column, and the unit is mm. Please use the most recent updated function named 
`gen_height_histogram_for_Roman(plydata, scanDirection, out_dir, sensor_d, center_position)`
This will consider a east or west side sensor, and add the newest origin offset in both side, and make the ground plane as hist[0], and use a 32 column definition.

With these numbers recorded, I would integrate all scanned result to create a full field highest and histogram file, using the according json file.

- And what for the 'hist'?
hist records the height distribution, 16 lengths of hist has the same definition as 'highest'. For 400 numbers in each element, it is bins to record how many points in that height level. you can treat col 0 as ground plane in the field, each height level equals to 1 cm in the field. So number(for example 5) in the 400 bin means there are 5 points in height level 400, which means 4 meters in the real world.

- Recommended way of summarizing data:
1. use 'gen_height_histogram_for_Roman' to create hist and highest array, use json file to determine the plot number for each bin, then change to whatever unit that might fit the database.
2. Notice that each plot might made up of several ply files, so you might create several different number for each plot for one day by a separated process.
3. create different quantiles data with 'hist' for each plot and use @rmgarnett 's formula to create an estimate height. reference codes: https://github.com/terraref/extractors-3dscanner/blob/master/plant_height/draw_field_scanned_in_grid.py#L195

**Comment by dlebauer** [comment]
Would it make sense to use the same workflow as we are using with other image data, i.e.:
* full field stitch
* plot subset 
* analysis, 

so that each result represents a single plot?

**Comment by max-zilla** [comment]
i believe we've discussed stitching point clouds into plots - I will try to track down comments from @solmazhajmohammadi

**Comment by dlebauer** [comment]
* I don't think that the files overlap (after R and L ply are merged into a single las)
* I think using the stitch then subset approach is okay. 

@solmazhajmohammadi can you confirm that independent passes do not overlap each other.

**Comment by dlebauer** [comment]
I think we can close this issue and open a new one with low priority that will deal with full field merge then split.

**Comment by solmazhajmohammadi** [comment]
@dlebauer to confirm the independencies between passes, we need to apply the transformation to all the ply files from a single day and merge them together. 
The transformation is available in [Issue 44](https://github.com/terraref/reference-data/issues/44)
